### PR TITLE
Integration xgboost issue2959

### PR DESCRIPTION
**Pull request by rfsiten**
[issue2959] Add xgboost handler and corresponding testcase.
###  ( **caution, this pr is based on [pr3435](https://github.com/mindsdb/mindsdb/pull/3435),** )

can specify the "task" argument in using block, options: auto, classifier, regressor, reg, cla, a, c, r
the default value is auto.

The handler will do:
1) Detect the task type (classify/regress) automatically;
2) Fill na in numeric feature column;
3) Drop id column automatically;
4) Encode categorical column by onehot/target-encoding;
5) While it's a classifier, val and val__prob will generated;
6) Use optuna (bayes based) to search best parameters.

examples:
```
-- xgboost regress
CREATE PREDICTOR titanic_regressor_2
FROM test_mindsdb_proxy
(select Sex, Fare, Pclass, Age from titanic_train_csv) 
PREDICT Age
USING engine='xgboost';

select t.Age
from xgboost.titanic_regressor_2 t
inner join test_mindsdb_proxy.titanic_train_csv p;


-- xgboost classify
CREATE PREDICTOR titanic_classifier_2
FROM test_mindsdb_proxy
(select * from titanic_train_csv) 
PREDICT Survived
USING engine='xgboost';

select t.Survived, t.Survived__prob
from xgboost.titanic_classifier_2 t
inner join test_mindsdb_proxy.titanic_train_csv p
where p.pclass = 1;

-- xgboost regress
CREATE PREDICTOR titanic_regressor_2
FROM test_mindsdb_proxy
(select Sex, Fare, Pclass, Age from titanic_train_csv) 
PREDICT Age
using engine='xgboost', automl=True, trail=30;

-- xgboost regress
CREATE PREDICTOR titanic_regressor_2
FROM test_mindsdb_proxy
(select Sex, Fare, Pclass, Age from titanic_train_csv) 
PREDICT Age
using engine='xgboost', task='regressor'; -- reg

-- xgboost classify
CREATE PREDICTOR titanic_classifier_2
FROM test_mindsdb_proxy
(select * from titanic_train_csv) 
PREDICT Survived
using engine='xgboost', automl=True;

-- xgboost classify
CREATE PREDICTOR titanic_classifier_2
FROM test_mindsdb_proxy
(select * from titanic_train_csv) 
PREDICT Survived
using engine='xgboost', task='cla';  -- classifier

```

hey, please review this pr, thank you so much. @paxcema @ZoranPandovski

### CODE REVIEW

**Comment by rfsiten** [comment]
Bayes based parameter search function added. In using block, user can start auto-parameter-tuning by automl=True, and set the number of trail points by trail=30.  More trail points set, more caculation will be done to find the best parameters, also more time will be cost.

**Comment by rfsiten** [comment]
hey, my friends @ZoranPandovski @paxcema @ea-rus , xgboost is ready !

File: mindsdb/integrations/handlers/xgboost_handler/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+mindsdb>=22.6.2.1
```
Comment (paxcema) [inline mindsdb/integrations/handlers/xgboost_handler/requirements.txt:1]: Update `mindsdb` and `mindsdb_sql` to latest versions :)

File: tests/unit/test_xgboost_handler.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import time
+from unittest.mock import patch
+import pandas as pd
+
+from mindsdb_sql import parse_sql
+
+# How to run:
+#  env PYTHONPATH=./ pytest tests/unit/test_xgboost_handler.py
+# Warning: uncomment and run too many sql-testcase in detector and forecaster will cost a long time
+from mindsdb.integrations.handlers.merlion_handler.adapters import DefaultForecasterAdapter, SarimaForecasterAdapter, \
```
Comment (paxcema) [inline tests/unit/test_xgboost_handler.py:11]: Leftovers from Merlion template? If so, remove please

File: tests/unit/test_xgboost_handler.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+from .executor_test_base import BaseExecutorTest
+
+import pandas as pd
+from pathlib import Path
+import requests
+
+
+class TestXgboost(BaseExecutorTest):
+    def get_titanic_train_df(self) -> pd.DataFrame:
+        df = pd.read_csv("/home/drill/Data/titanic/train.csv")
```
Comment (paxcema) [inline tests/unit/test_xgboost_handler.py]: Please change the .csv path to one inside the actual repo (upload a small file if necessary), or maybe to one of our publicly hosted datasets like "Home Rentals"

File: tests/unit/test_xgboost_handler.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        df_train = self.get_titanic_train_df()
+        df_test = self.get_titanic_test_df()
+        self.set_handler(mock_handler, name='pg', tables={'titanic_train': df_train, "titanic_test": df_test})
+        self.exec_train_and_predict(mock_handler=mock_handler, model_name="classifier", predict_val="Survived")
+
+    def exec_train_and_predict(self, mock_handler, model_name, predict_val, is_classifier=True):
+        # create predictor
+        train_cols = "PassengerId, Pclass, Sex, Age, SibSp, Parch, Fare, Cabin, Embarked"
+        if is_classifier:
+            train_cols = "PassengerId, Pclass, Sex, Age, SibSp, Parch, Fare, Cabin, Embarked, Survived"
```
Comment (paxcema) [inline tests/unit/test_xgboost_handler.py:57]: Alternatively, `train_cols += ', Survived'`

File: mindsdb/integrations/handlers/xgboost_handler/xgboost_handler.py [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+
+
+def enum_to_str(type_class: Enum) -> str:
+    all = []
+    for element in type_class:
+        all.append(element.name)
+    return "|".join(all)
+
+
+def to_ts_dataframe(df: pd.DataFrame, time_col=None) -> str:
```
Comment (paxcema) [inline mindsdb/integrations/handlers/xgboost_handler/xgboost_handler.py:48]: From what I can see, this handler does not support time series in its current form. Do we need this method, then?

File: mindsdb/integrations/handlers/xgboost_handler/xgboost_handler.py [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+                or target_type is np.dtype("float") or target_type is np.dtype("int"):
+            if label_cnt <= UNIQUE_CNT_TO_BE_REG:
+                return TaskType.classifier, label_cnt
+            else:
+                return TaskType.regressor, 0
+        else:
+            return TaskType.classifier, label_cnt
+
+
+class FeatureProcessor:
```
Comment (paxcema) [inline mindsdb/integrations/handlers/xgboost_handler/xgboost_handler.py:96]: Just FYI, we are working on a package (`dataprep_ml`) that will be handler-agnostic and may help here once it's fully released.

Comment (paxcema) [reply mindsdb/integrations/handlers/xgboost_handler/xgboost_handler.py:96]: We could add support for `category_encoders` there, but this is fine for now.

File: mindsdb/integrations/handlers/xgboost_handler/xgboost_handler.py [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+        feature_transformer_bytes = pickle.dumps(feature_transformer)
+        self.model_storage.json_set(self.PERSISIT_ARGS_KEY_IN_JSON_STORAGE, args)
+        self.model_storage.file_set(self.PERSISIT_MODEL_FILE_NAME, model_bytes)
+        self.model_storage.file_set(self.PERSISIT_FEATURE_TRANSFORMER, feature_transformer_bytes)
+        log.logger.info(F"Model saved.")
+
+    def predict(self, df, args):
+        # read from storage
+        args = self.model_storage.json_get(self.PERSISIT_ARGS_KEY_IN_JSON_STORAGE)
+        feature_transofrmer_bytes = self.model_storage.file_get(self.PERSISIT_FEATURE_TRANSFORMER)
```
Comment (paxcema) [inline mindsdb/integrations/handlers/xgboost_handler/xgboost_handler.py:367]: typo in variable name

File: mindsdb/integrations/handlers/xgboost_handler/xgboost_handler.py [hunk 8]
```diff
@@ -0,0 +1,471 @@
...
+        pred_data = boost_model.predict(p_feature_data)
+        if task_type == TaskType.classifier:
+            label_arr = feature_transformer.label_encoder.inverse_transform([i for i in range(class_num)])
+            pred_label = [str(label_arr[i]) for i in np.argmax(pred_data, axis=1)]
+            pred_probs_arr = []
+            for probs in pred_data:
+                prob_items = []
+                for i, prob in enumerate(probs):
+                    prob_items.append("\"" + str(label_arr[i]) + "\"=" + str(round(prob, 5)))
+                pred_probs_arr.append("{" + ",".join(prob_items) + "}")
```
Comment (paxcema) [inline mindsdb/integrations/handlers/xgboost_handler/xgboost_handler.py:386]: Why is `pred_probs_arr` a list of strings? I would imagine a list of lists (with `object` dtype) to be okay for the MindsDB planner + executioner (we do this in the Lightwood handler).

**Review by paxcema — CHANGES_REQUESTED** [review]
Great job @rfsiten, in general this is looking good.

I do have one major concern, which is that this introduces a big chunk of automated (and opinionated) feature engineering that we should be testing very well before declaring this as a stable backend. The handler will be experimental for now, so this is fine, but we definitely want to ensure a good test coverage for all the new logic in here in a new PR.

Additionally, some components here (e.g. data type inference and parts of the feature engineering like imputing) will have overlap with new packages we are developing, at which point it may make sense to merge some of the work you've done here into them, so that all other handlers can leverage it. I'll let you know when the time comes :)

Finally, please resolve the merge conflict so that we can try to run CI on this.

**Comment by paxcema** [comment]
Oh! And I forgot: please add docstrings in the main components, explaining what they do, what can be customized through `using`, and how they relate to the rest of the handler.

**Comment by rfsiten** [comment]
@paxcema hey, I will resolve the problems in 1~2 weeks.

**Comment by ZoranPandovski** [comment]
@rfsiten Do you have any progress here, or do you need help from us?

File: mindsdb/integrations/handlers/xgboost_handler/__init__.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+    from .xgboost_handler import XgboostHandler as Handler
+    import_error = None
+except Exception as e:
+    Handler = None
+    import_error = e
+
+title = 'Xgboost'
+name = 'xgboost'
+type = HANDLER_TYPE.ML
+permanent = False
```
Comment (ea-rus) [inline mindsdb/integrations/handlers/xgboost_handler/__init__.py:14]: from examples it looks like it should be permanent integration (no need to create xgboost ml engine)
` permanent = False`

**Comment by paxcema** [comment]
@rfsiten Hi Franky, do you expect to be able to continue working on this in the short term? If not, please let us know. Thanks!

**Comment by ZoranPandovski** [comment]
@rfsiten Once you have time to make the changes, please re-open the PR

**Pull request closed without merging**